File: nodejs/src/workerCausality.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

function object(value: unknown): value is Record<string, unknown> {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}

function fields(value: Record<string, unknown>, allowed: readonly string[]): boolean {
    return Object.keys(value).every((key) => allowed.includes(key));
}

function text(value: unknown): value is string {
    return typeof value === "string" && value.length > 0 && Buffer.byteLength(value, "utf8") <= 256;
}

function small(value: unknown, budget: { remaining: number }, depth = 0): boolean {
    budget.remaining--;
    if (budget.remaining < 0 || depth > 32) return false;
    if (typeof value === "string") budget.remaining -= value.length;
    if (object(value)) {
        for (const [key, item] of Object.entries(value)) {
            if (!small(key, budget, depth + 1) || !small(item, budget, depth + 1)) return false;
        }
    } else if (Array.isArray(value)) {
        for (const item of value) if (!small(item, budget, depth + 1)) return false;
    }
    return budget.remaining >= 0;
}

function reference(value: unknown, type: string): boolean {
    return (
        object(value) &&
        fields(value, ["sessionId", "eventId", "agentId", "eventType", "provenance"]) &&
        text(value.sessionId) &&
        /^[A-Za-z0-9_-]+$/.test(value.sessionId) &&
        typeof value.eventId === "string" &&
        /^[\da-f]{8}-[\da-f]{4}-[\da-f]{4}-[\da-f]{4}-[\da-f]{12}$/i.test(value.eventId) &&
        (!("agentId" in value) || text(value.agentId)) &&
        value.eventType === type &&
        (value.provenance === "native" || value.provenance === "ahp_coordinator")
    );
}

function source(value: unknown): boolean {
    if (!object(value) || !object(value.input)) return false;
    const input = value.input;
    return (
        fields(value, [
            "input",
            "admissions",
            "captureComplete",
            "completion",
            "notification",
            "admittedDuring",
        ]) &&
        fields(input, ["queueItemId", "agentId", "sender", "senderBridges"]) &&
        typeof value.captureComplete === "boolean" &&
        typeof input.queueItemId === "string" &&
        /^[\da-f]{8}-[\da-f]{4}-[\da-f]{4}-[\da-f]{4}-[\da-f]{12}$/i.test(input.queueItemId) &&
        text(input.agentId) &&
        (!("sender" in input) || reference(input.sender, "tool.execution_start")) &&
        (!("senderBridges" in input) ||
            ("sender" in input &&
                Array.isArray(input.senderBridges) &&
                input.senderBridges.length <= 32 &&
                input.senderBridges.every(
                    (edge: unknown) =>
                        object(edge) &&
                        fields(edge, ["source", "reported"]) &&
                        reference(edge.source, "tool.execution_start") &&
                        reference(edge.reported, "tool.execution_start")
                ))) &&
        Array.isArray(value.admissions) &&
        value.admissions.length <= 32 &&
        value.admissions.every(
            (item: unknown) =>
                object(item) &&
                fields(item, ["kind", "messageId", "event", "ahpTurnId"]) &&
                (item.kind === "queued_input" || item.kind === "system_continuation") &&
                text(item.messageId) &&
                (!("event" in item) ||
                    (reference(item.event, "user.message") &&
                        (item.event as Record<string, unknown>).agentId === input.agentId)) &&
                (!("ahpTurnId" in item) ||
                    (typeof item.ahpTurnId === "string" &&
                        /^[\da-f]{8}-[\da-f]{4}-[\da-f]{4}-[\da-f]{4}-[\da-f]{12}$/i.test(
                            item.ahpTurnId
                        )))
        ) &&
        (!("completion" in value) || reference(value.completion, "subagent.completed")) &&
        (!("admittedDuring" in value) || reference(value.admittedDuring, "assistant.turn_start")) &&
        (!("notification" in value) ||
            (object(value.notification) &&
                fields(value.notification, ["deliveryId", "event", "mode"]) &&
                typeof value.notification.deliveryId === "string" &&
                /^[\da-f]{8}-[\da-f]{4}-[\da-f]{4}-[\da-f]{4}-[\da-f]{12}$/i.test(
                    value.notification.deliveryId
                ) &&
                (value.notification.mode === "queued" || value.notification.mode === "immediate") &&
                (!("event" in value.notification) ||
                    reference(value.notification.event, "system.notification"))))
    );
}

/** Decode availability only; consumers must still validate placement and enclosing self identity. */
export function withWorkerCausality<T extends object>(data: T): T {
    if (!("workerCausality" in data)) return data;
    const value = data.workerCausality;
    const valid =
        object(value) &&
        fields(value, ["version", "observationProvenance", "sources", "captureComplete"]) &&
        value.version === 1 &&
        (value.observationProvenance === "native" ||
            value.observationProvenance === "ahp_coordinator") &&
        typeof value.captureComplete === "boolean" &&
        Array.isArray(value.sources) &&
        value.sources.length <= 32 &&
        small(value, { remaining: 4096 }) &&
        value.sources.every(
            (item: unknown) =>
                source(item) &&
                (!value.captureComplete ||
                    (item as Record<string, unknown>).captureComplete === true)
        ) &&
        Buffer.byteLength(JSON.stringify({ workerCausality: value }), "utf8") <= 4096;
    if (valid) return data;
    console.warn("Ignoring invalid, unsupported or oversized workerCausality metadata");
    const copy = { ...data };
    delete (copy as { workerCausality?: unknown }).workerCausality;
    return copy;
}

/** Apply the same diagnostic tolerance to historical events as live delivery. */
export function withWorkerCausalityEvents<T extends { events: { data: object }[] }>(result: T): T {
    const events = result.events.map((event) => {
        const data = withWorkerCausality(event.data);
        return data === event.data ? event : { ...event, data };
    });
    return events.every((event, index) => event === result.events[index])
        ? result
        : { ...result, events };
}

File: rust/src/worker_causality.rs
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

use std::io::{self, Write};

use serde::de::DeserializeOwned;
use serde::{Deserialize, Deserializer};
use serde_json::Value;

struct Budget(usize);

impl Write for Budget {
    fn write(&mut self, bytes: &[u8]) -> io::Result<usize> {
        self.0 = self.0.saturating_add(bytes.len()).min(4097);
        if self.0 > 4096 {
            Err(io::Error::other("optional worker metadata budget"))
        } else {
            Ok(bytes.len())
        }
    }

    fn flush(&mut self) -> io::Result<()> {
        Ok(())
    }
}

fn text(value: &Value) -> bool {
    value.as_str().is_some_and(|value| {
        !value.is_empty() && value.len() <= 256 && !value.chars().any(char::is_control)
    })
}

fn uuid(value: &Value) -> bool {
    value.as_str().is_some_and(|value| {
        value.len() == 36
            && value.bytes().enumerate().all(|(index, byte)| {
                if matches!(index, 8 | 13 | 18 | 23) {
                    byte == b'-'
                } else {
                    byte.is_ascii_hexdigit()
                }
            })
    })
}

fn fields(value: &Value, allowed: &[&str]) -> bool {
    value
        .as_object()
        .is_some_and(|object| object.keys().all(|key| allowed.contains(&key.as_str())))
}

fn reference(value: &Value, event_type: &str) -> bool {
    fields(
        value,
        &["sessionId", "eventId", "agentId", "eventType", "provenance"],
    ) && text(&value["sessionId"])
        && uuid(&value["eventId"])
        && value.get("agentId").is_none_or(text)
        && value["eventType"] == event_type
        && matches!(
            value["provenance"].as_str(),
            Some("native" | "ahp_coordinator")
        )
}

fn source(value: &Value) -> bool {
    let input = &value["input"];
    let Some(admissions) = value["admissions"].as_array() else {
        return false;
    };
    if !fields(
        value,
        &[
            "input",
            "admissions",
            "captureComplete",
            "completion",
            "notification",
            "admittedDuring",
        ],
    ) || !fields(
        input,
        &["queueItemId", "agentId", "sender", "senderBridges"],
    ) || !value["captureComplete"].is_boolean()
        || !uuid(&input["queueItemId"])
        || !text(&input["agentId"])
        || admissions.len() > 32
        || input
            .get("sender")
            .is_some_and(|sender| !reference(sender, "tool.execution_start"))
    {
        return false;
    }
    if let Some(edges) = input.get("senderBridges") {
        let Some(edges) = edges.as_array() else {
            return false;
        };
        if input.get("sender").is_none()
            || edges.len() > 32
            || edges.iter().any(|edge| {
                !fields(edge, &["source", "reported"])
                    || !reference(&edge["source"], "tool.execution_start")
                    || !reference(&edge["reported"], "tool.execution_start")
            })
        {
            return false;
        }
    }
    if admissions.iter().any(|admission| {
        !fields(admission, &["kind", "messageId", "event", "ahpTurnId"])
            || !matches!(
                admission["kind"].as_str(),
                Some("queued_input" | "system_continuation")
            )
            || !text(&admission["messageId"])
            || admission.get("ahpTurnId").is_some_and(|turn| !uuid(turn))
            || admission.get("event").is_some_and(|event| {
                !reference(event, "user.message") || event["agentId"] != input["agentId"]
            })
    }) {
        return false;
    }
    for (field, event_type) in [
        ("completion", "subagent.completed"),
        ("admittedDuring", "assistant.turn_start"),
    ] {
        if value
            .get(field)
            .is_some_and(|event| !reference(event, event_type))
        {
            return false;
        }
    }
    value.get("notification").is_none_or(|notification| {
        fields(notification, &["deliveryId", "event", "mode"])
            && uuid(&notification["deliveryId"])
            && matches!(notification["mode"].as_str(), Some("queued" | "immediate"))
            && notification
                .get("event")
                .is_none_or(|event| reference(event, "system.notification"))
    })
}

// Availability only: consumers must additionally validate placement and enclosing self identity.
fn supported(value: &Value) -> bool {
    fields(
        value,
        &[
            "version",
            "observationProvenance",
            "sources",
            "captureComplete",
        ],
    ) && value["version"].as_u64() == Some(1)
        && matches!(
            value["observationProvenance"].as_str(),
            Some("native" | "ahp_coordinator")
        )
        && value["captureComplete"].is_boolean()
        && value["sources"].as_array().is_some_and(|sources| {
            sources.len() <= 32
                && sources.iter().all(|item| {
                    source(item)
                        && (value["captureComplete"] != true || item["captureComplete"] == true)
                })
        })
}

pub(crate) fn deserialize_optional<'de, D: Deserializer<'de>, T: DeserializeOwned>(
    deserializer: D,
) -> Result<Option<T>, D::Error> {
    let value = Value::deserialize(deserializer)?;
    let field = std::collections::BTreeMap::from([("workerCausality", &value)]);
    if serde_json::to_writer(&mut Budget(0), &field).is_ok()
        && supported(&value)
        && let Ok(capture) = serde_json::from_value(value)
    {
        return Ok(Some(capture));
    }
    tracing::warn!("Ignoring invalid, unsupported or oversized workerCausality metadata");
    Ok(None)
}
